draft-ietf-protected-audience-key-value-services.md

```diff
@@ -318,9 +318,43 @@ Each key group is expected to have exactly one tag from the following list:
 | renderUrls | "renderUrls" represent URLs for advertisements to be looked up from the service's internal datastore. |
 | adComponentRenderUrls | "adComponentRenderUrls" represent component URLs for advertisements to be looked up from the service's internal datastore. |
 
-### Generating a Request
-
-TODO
+### Generating a Request {#request-generate}
+
+This section describes how the client MAY form and serialize request messages in order to fetch values from the Trusted Key Value server.
+
+This algorithm takes as input an [HPKE] `public key` and its associated `key id`, a `metadata` map for global configuration, where both keys and values are strings, and `compression groups`, which is a list of compression group structs. Each struct contains a compression group with a compression group ID and one or more partitions. Each partition includes multiple keys, such as "id," whose value is an integer; "metadata," whose value is a map where both keys and values are strings; and "namespace," whose value is a map where keys are strings and values are lists of strings.
+
+The output is an [HPKE] ciphertext encrypted `request` and a context `request context`.
+
+1. Let `request map` be an empty map.
+1. Let `partitions` be an empty array.
+1. For each `group` in `compression groups`:
+  1. For each `partition` in `group["partitions"]:
+    1. Let `p` be an empty map.
+    1. Set `p["compressionGroupId"]` to `group[compressionGroupId"]`.
+    1. Set `p["id"]` to `partition["id"]`.
+    1. Set `p["metadata"]` to `partition["metadata"].
+    1. Let `arguments` be an empty array.
+    1. For each `tag` → `value` in `partition["namespace"]`:
```

Comment:
We better not copy each tag blindly. We should look through each tag in [2.2.3.1](https://privacysandbox.github.io/draft-ietf-protected-audience-key-value-service/draft-ietf-protected-audience-key-value-services.html#section-2.2.3.1) so that we only extract recognized tags. Otherwise it may be a problem if we add new tag types.

Reply:
I still think it is cleaner to loop through the available tags and then see if the tag exists, but I guess this works as well.